### DISCUSSION TITLE
bz error: Secular determinant generated by symmetry operations may be faulty.

### DISCUSSION THREAD

**Discussion by kamischi** [openmopac/mopac#187] [Q&A]
I perform a PM7 calculation on a 2D periodic honeycomb system of Sc2O3. The actual calculation runs nicely without any obvious faults. But when I try to get the band structure with "bz", I get this error:

Secular determinant generated by symmetry operations may be faulty.

As suggested from the output I tried the options SYMTRZ and ROTFOK in bz. This produced some more output, but it did not really give me a better idea of what is wrong.

I have already tried NOSYM as keyword in mopac and switched from internal coordinates to xyz, but none of this made any change. So far, I did quite a number of similar calculations with a 2D honeycomb system with Ti2O3 and never had this problem. 

Does anyone have any ideas?

Input/Output files are appended as zip archive.

Best wishes - Michael.

[Sc2O3 6-Ring.zip](https://github.com/openmopac/mopac/files/13729446/Sc2O3.6-Ring.zip)

**Comment by godotalgorithm** [comment]
That error is triggered by a numerical threshold that I don't understand at first glance. I'll bring this up with Jimmy Stewart during our next meeting in the new year, and I'll run some tests to see why some calculations aren't triggering this threshold and others are. It might simply be that this threshold was given too large a value and is producing false positives in the test for an error.

**Comment by kamischi** [comment]
Here is the same structure of Sc2O3, but a different definition of it. With this definition, the .brz file does not lead to the error with bz. Maybe it helps.
If the [5x5] structure is too large to debug, I could also try, if the same problem shows up with the [3x3] structure.

[Sc2O3 6-ring Working.zip](https://github.com/openmopac/mopac/files/13752583/Sc2O3.6-ring.Working.zip)

Happy holidays and New Year - Michael.

**Comment by godotalgorithm** [comment]
I've tentatively fixed this bug in the development version of MOPAC, but I haven't been able to test out my fix because the pre-compiled Windows version of BZ isn't working at all for me right now. How are you running BZ? Are you compiling some modified version without the QuickWin-based graphics?

I might just remove the interactive mode and graphics from BZ and require the use of a .dat input file, which would allow BZ to be run on Mac and Linux. Have you used the interactive mode of BZ at all, and if so did you find it useful at all?

**Reply by godotalgorithm** [reply]
BZ is compiling and running again in the development version, but unfortunately my bug fix for BZ hasn't completely fixed this problem. I've reopened the issue, and I should be able to fix this more definitively now that I can compile a functional BZ program for testing.

**Comment by kamischi** [comment]
Hi

Sorry for the late answer. I actually run the windows BZ binary on our windows server 2019. Initially I used the interactive mode, mainly to find out the parameters, like values of k. Once I had them, I created a command file for convenience as well as documentation. Once I had a plausible set of input values, it was very useful to create bandstructure plots. My system is similar to graphene und only the band structure plot shows if the Dirac cones are present at the Fermi edge or not.

I had an initial try to use SDL or similar for a linux/macOS version, but did not really get far.

Michael.

**Reply by godotalgorithm** [reply]
Ok, it also looks like you are running an older version of MOPAC, and you'll have to update to the latest version once I do get the BZ bug fixed. My tentative plan is to fix the bug as soon as I can and make a patch-level release of MOPAC with a working version of BZ (it's been broken for over a year in the releases, and no one has ever complained about it because BZ is not frequently used). After that, I'll probably reorganize BZ a bit to isolate the graphical components so that I can compile a non-graphical version for Mac and Linux, which will be the tentative target of the next minor-version release.

What happens to BZ after that will depend on user and developer interest - I might eventually strip out all of the graphics for uniformity over operating systems, or commit to some graphics solution. A possible intermediate solution is to refactor BZ into a Python program with a Fortran library. Python would handle the graphics and interactivity through [MatPlotLib's interactive mode](https://matplotlib.org/stable/users/explain/figure/interactive.html), and I would retain all of the computational components of the Fortran code. This solution would require users to have Python installed, but MOPAC is already moving in a direction of increased Python integration (the testing system already requires Python, and I should be finally releasing an official Python interface to MOPAC this year).

**Comment by kamischi** [comment]
Clarification note: I run the underlying MOPAC calculations with the current version on macOS. Re BZ, I use the old version on a windows server.

**Comment by godotalgorithm** [comment]
A verified bug fix is now merged into the development version. I have noticed from your examples that debug keywords passed into BZ are preventing it from running normally in addition to printing out debug information. I'm going to try to fix this behavior before I release the patch.

**Reply by godotalgorithm** [reply]
Actually, the behavior of BZ is fine, just slightly inconsistent from MOPAC and PARAM. It reads keywords from the same line as the integer specifying the mode (point/line/surface). In the first example you sent me, you had the keywords on a separate line in the `.dat` file, which triggered debugging information but not the actual band-structure calculations. Also, all debug information is printed to a common `.txt` file that is overwritten between interactive steps. I may alter this behavior in the future, but it does appear to work as intended. The intent is just a bit inconvenient.

**Reply by godotalgorithm** [reply]
Ok, the patched version is released now (v22.1.1). The updated version of BZ should resolve your original problem, and let me know if it doesn't, so that I can take further action.
